pixeltable/exprs/expr.py

```diff
@@ -388,6 +390,36 @@ def from_dict_list(cls, dict_list: List[Dict]) -> List[Expr]:
     def _from_dict(cls, d: Dict, components: List[Expr]) -> Expr:
         assert False, 'not implemented'
 
+    def astype(self, new_type: ts.ColumnType) -> 'pixeltable.exprs.TypeCast':
+        from pixeltable.exprs import TypeCast
+        return TypeCast(self, new_type)
+
+    def apply(self, fn: Callable, *, col_type: Optional[ts.ColumnType] = None) -> 'pixeltable.exprs.FunctionCall':
+
+        if col_type is not None:
+            # col_type is specified explicitly
+            fn_type = col_type
+        elif fn in _known_applicator_types:
+            # For convenience, various built-ins and other Python functions that don't
+            # have type hints are hardcoded
+            fn_type = _known_applicator_types[fn]
+        elif 'return_type' in typing.get_type_hints(fn):
+            # Attempt to infer the column type from the return type of the callable;
+            # this will set fn_type to None if it cannot be inferred
+            return_type = typing.get_type_hints(fn)['return_type']
+            fn_type = ts.ColumnType.infer_literal_type(return_type)
+        else:
+            # No type hint
+            fn_type = None
+
+        if fn_type is None:
+            raise RuntimeError(
+                f'Column type of `{fn.__name__}` cannot be inferred. Use `.apply({fn.__name__}, col_type=...)` to specify.'
+            )
+
+        function = make_function(return_type=fn_type, param_types=[self.col_type], eval_fn=lambda x: fn(x))
```

Comment:
This is missing a bunch of verification/other logic, such as
- does this have exactly one non-default parameter?
- deal with this being a keyword-only parameter
- if fn has type annotations, do type checking

Reply:
I added some validation for parameter counts. I'm going to pass on doing type checking for now (we can discuss in person).